Automattic/jetpack
# Connection package: pass along redirectUri to connectUser

### Summary




The Jetpack Partner Coupon redeem flow shows the `<ConnectScreen>` component for sites that has been registered but doesn't have a connected user (for any reason), which isn't a "normal" scenario we show the connect screen in.

_Normally when a site has been registered, we no longer show `<ConnectScreen>` but instead relies on other prompts to connect the user if needed._

So when we [on December 17th merged a refactor to extract connection logic to custom hooks](https://href.li/?https://github.com/Automattic/jetpack/pull/21885), we never tested for the `redirectUri` to be passed along to `connectUser` for "userless connections".
The `<ConnectScreen>` used to work before the refactor because we used `<ConnectUser>` which correctly passes the redirect along to `restApi.fetchAuthorizationUrl()`.

You can read more in pdpAdu-al-p2 but the TL;DR is that we no longer pass along the partner coupon to Calypso which means users cannot redeem their coupon.

### Changes proposed in this Pull Request:

* Allow for a `redirectUri` to be passed to `connectUser` action (like how the `<ConnectUser>` component passes it along)
* 👆  will indirectly allow for the `redirectUri` value in `<ConnectScreen>` to be passed along to `connectUser`.

### Jetpack product discussion



* pdpAdu-al-p2

### Does this pull request change what data or activity we track or use?



It doesn't introduce any new events; nor should it change the behaviour of any existing tracking activities.

### Testing instructions:






* Spin up this branch either locally or as a JN beta branch site
* Go to `/wp-admin/admin.php?jetpack-partner-coupon={COUPON}`
  * Replace `{COUPON}` with the coupon found here: 2bc89-pb
* Press `Set up & redeem Jetpack Backup` (see screenshot below)
* You should now end up on WPCOM
* Do not Authorize your WPCOM account, but instead go back to `/wp-admin/admin.php?page=jetpack&showCouponRedemption=1`
* You should now see the component with `Set up & redeem Jetpack Backup` again (see screenshot below)
* Press the `Set up & redeem Jetpack Backup` button and `Authroize` the on WPCOM.
* You should now end up in checkout with a 100% discounted price coupon applied (do not complete checkout since others might want to test with it as well). Reaching checkout is already proof that it works.

_**Bonus:** feel free to go through the same steps as above, but with Jetpack 10.5+ to see it fail._

**Screenshot of the referenced partner coupon redeem component**

![Screenshot 2022-02-24 at 18 13 38](https://user-images.githubusercontent.com/3846700/155573699-7ca711b7-a5f9-4f17-86ee-9f49b83602ba.png)
